tests/components/sensor/test_awair.py

```diff
@@ -0,0 +1,249 @@
+    assert sensor.state == '22.4'
+    assert entity.device_class == DEVICE_CLASS_TEMPERATURE
+    assert entity.unit_of_measurement == TEMP_CELSIUS
+
+
+async def test_awair_humid(hass):
+    """Test that we create a humidity sensor."""
+    await setup_awair(hass)
+
+    sensor = hass.states.get('sensor.awair_humidity')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_humidity')
+    assert sensor.state == '32.73'
+    assert entity.device_class == DEVICE_CLASS_HUMIDITY
+    assert entity.unit_of_measurement == '%'
+
+
+async def test_awair_co2(hass):
+    """Test that we create a CO2 sensor."""
+    await setup_awair(hass)
+
+    sensor = hass.states.get('sensor.awair_co2')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_co2')
+    assert sensor.state == '612'
+    assert entity.device_class == DEVICE_CLASS_CARBON_DIOXIDE
+    assert entity.unit_of_measurement == 'ppm'
+
+
+async def test_awair_voc(hass):
+    """Test that we create a CO2 sensor."""
+    await setup_awair(hass)
+
+    sensor = hass.states.get('sensor.awair_voc')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_voc')
+    assert sensor.state == '1012'
+    assert entity.device_class == DEVICE_CLASS_VOLATILE_ORGANIC_COMPOUNDS
+    assert entity.unit_of_measurement == 'ppb'
+
+
+async def test_awair_dust(hass):
+    """Test that we create a pm25 sensor."""
+    await setup_awair(hass)
+
+    # The Awair Gen1 that we mock actually returns 'DUST', but that
+    # is mapped to pm25 internally so that it shows up in Homekit
+    sensor = hass.states.get('sensor.awair_pm25')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_pm25')
+    assert sensor.state == '6.2'
+    assert entity.device_class == DEVICE_CLASS_PM2_5
+    assert entity.unit_of_measurement == 'µg/m3'
+
+
+async def test_awair_unsupported_sensors(hass):
+    """Ensure we don't create sensors the stubbed device doesn't support."""
+    await setup_awair(hass)
+
+    # Our tests mock an Awair Gen 1 device, which should never return
+    # PM10 sensor readings. Assert that we didn't create a pm10 sensor,
+    # which could happen if someone were ever to refactor incorrectly.
+    assert hass.states.get('sensor.awair_pm10') is None
+
+
+async def test_async_update(hass):
+    """Ensure we can update sensors."""
+    await setup_awair(hass)
+
+    fixture = json.loads(load_fixture('awair_air_data_latest_updated.json'))
+
+    with patch('python_awair.AwairClient.air_data_latest',
+               return_value=mock_coro(fixture)):
+        entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_score')
+
+        # pylint: disable=protected-access
+        await entity._data.async_update(no_throttle=True)
+
+        # these sensors are marked as should_poll, so force the interval
+        for sensor in hass.states.async_all():
+            e_id = sensor.entity_id
+            await hass.helpers.entity_component.async_update_entity(e_id)
```

Comment:
Instead patch `homeassistant.util.dt.utcnow` and use `async_fire_time_changed` callback from tests/common.py to fire a mock time event. Eg like this:
https://github.com/home-assistant/home-assistant/blob/dev/tests/components/alarm_control_panel/test_manual.py#L78-L82

Make sure to patch time when setting up the component too, since that's when the first update happens.

Reply:
:+1: I like `async_fire_time_changed`, that's elegant.